MH-12641 Asset manager conflict checks are very slow

krinnewitz opened the pull request:
When scheduling new events for a certain capture agent, the
scheduler service checks for conflicts with existing events already
scheduled on that capture agent.

For this check, queries are created using the asset manager's query
builder. Unfortunately, due to the nature of the underlying database
schema, the resulting SQL queries contain many nested queries and hence,
are very slow if there already is a higher number of scheduled events.

The best approach to solve this would be to change the database schema.
The problem is mainly caused by the concept of generic 'properties'
which are related to snapshots and are managed by the asset manager.
However, changing the schema would mean that we need a database
migration and it would yield architectural issues since the separation
between asset manager and scheduler service would break.

Since there currently is not enough time to change the database schema,
this patch chooses a different approach. The measurements include:

- Don't do three queries for each event to schedule, but only one query
  for all the events to schedule. When scheduling 50 events, we now use
  one query instead of 150 queries.
- Use a simpler query with less nested queries.

This is achieved by reading all events for the capture agent in
question and then doing the conflict checks in memory. Of course, there
may be lots of events and performance issues can still arise when
reading so many rows from the database. Also, no limit is used which
generally is bad practice. In the worst case, opencast could run out of
memory or reading all the rows from the database could take forever.
However, productive systems are expected to have a moderate number of
scheduled recordings per capture agent which should not cause the
problems mentioned.

Disclaimer: This solution is not optimal since it utilizes an anti
pattern, but it solves the performance issue for now. It should be
considered as a temporary workaround which as well has its problems.

To test this, you first have to schedule some events and then try to schedule additional events which may or may not cause conflicts.

This work is sponsored by SWITCH.

Commit:
MH-12641 Asset manager conflict checks are very slow

When scheduling new events for a certain capture agent, the
scheduler service checks for conflicts with existing events already
scheduled on that capture agent.

For this check, queries are created using the asset manager's query
builder. Unfortunately, due to the nature of the underlying database
schema, the resulting SQL queries contain many nested queries and hence,
are very slow if there already is a higher number of scheduled events.

The best approach to solve this would be to change the database schema.
The problem is mainly caused by the concept of generic 'properties'
which are related to snapshots and are managed by the asset manager.
However, changing the schema would mean that we need a database
migration and it would yield architectural issues since the separation
between asset manager and scheduler service would break.

Since there currently is not enough time to change the database schema,
this patch chooses a different approach. The measurements include:

- Don't do three queries for each event to schedule, but only one query
  for all the events to schedule. When scheduling 50 events, we now use
  one query instead of 150 queries.
- Use a simpler query with less nested queries.

This is achieved by reading all events for the capture agent in
question and then doing the conflict checks in memory. Of course, there
may be lots of events and performance issues can still arise when
reading so many rows from the database. Also, no limit is used which
generally is bad practice. In the worst case, opencast could run out of
memory or reading all the rows from the database could take forever.
However, productive systems are expected to have a moderate number of
scheduled recordings per capture agent which should not cause the
problems mentioned.

Disclaimer: This solution is not optimal since it utilizes an anti
pattern, but it solves the performance issue for now. It should be
considered as a temporary workaround which as well has its problems.

smarquard commented:
To quantify the expected number of scheduled recordings per Capture Agent, in 2017 our busiest venue had just over 500 recordings across the year, and a theoretical upper limit (for us) is probably about 1200 recordings per CA per year (i.e. scheduled at any one time).

krinnewitz commented:
Sounds good. I tested locally with 700 scheduled recordings and on my machine, the request to /conflict took about 20 seconds.

staubesv commented:
@smarquard What performance did 3.x deliver for the numbers you mention?

@krinnewitz Will the scheduled event become an 'non-scheduled' events once being processed?

krinnewitz commented:
@staubesv At least there is code in the scheduler service which seems to do that. However, I haven't checked under which circumstances this code is called.

smarquard commented:
The query that fetches the scheduled events for a capture agents looks like this (from the mysql slow query log). Note the relatively large Rows_examined in comparison to Rows_sent.

```
# Time: 2018-01-17T15:24:20.286898Z
# User@Host: opencast[opencast] @ devslsmth001.uct.ac.za [137.158.158.193]  Id:   896
# Query_time: 3.428755  Lock_time: 0.000803 Rows_sent: 436  Rows_examined: 155287
SET timestamp=1516202660;
SELECT t0.id, t0.archival_date, t0.availability, t0.mediapackage_id, t0.mediapackage_xml, t0.organization_id, t0.owner, t0.series_id, t0.version, t1.id, t1.val_bool, t1.val_date, t1.val_long, t1.mediapackage_id, t1.namespace, t1.property_name, t1.val_string FROM mh_assets_snapshot t0 LEFT OUTER JOIN mh_assets_properties t1 ON (((t1.mediapackage_id = t0.mediapackage_id) AND ((t1.namespace = 'org.opencastproject.scheduler') AND (t1.property_name = 'start'))) OR ((t1.mediapackage_id = t0.mediapackage_id) AND ((t1.namespace = 'org.opencastproject.scheduler') AND (t1.property_name = 'end')))) WHERE ((((t0.organization_id = 'mh_default_org') AND EXISTS (SELECT 1 FROM mh_assets_properties t2 WHERE ((t0.mediapackage_id = t2.mediapackage_id) AND (t2.namespace = 'org.opencastproject.scheduler'))) ) AND ((NOT EXISTS (SELECT 1 FROM mh_assets_properties t3 WHERE (((t0.mediapackage_id = t3.mediapackage_id) AND (t3.namespace = 'org.opencastproject.scheduler.trx')) AND (t3.property_name = 'transaction_id')))  AND (t0.version = (SELECT MAX(t4.version) FROM mh_assets_snapshot t4 WHERE (t4.mediapackage_id = t0.mediapackage_id)))) OR (EXISTS (SELECT 1 FROM mh_assets_properties t5 WHERE (((t0.mediapackage_id = t5.mediapackage_id) AND (t5.namespace = 'org.opencastproject.scheduler.trx')) AND (t5.property_name = 'transaction_id')))  AND EXISTS (SELECT 1 FROM mh_assets_properties t6 WHERE ((((t0.mediapackage_id = t6.mediapackage_id) AND (t6.namespace = 'org.opencastproject.scheduler')) AND (t6.property_name = 'version')) AND (t0.version = t6.val_long))) ))) AND EXISTS (SELECT 1 FROM mh_assets_properties t7 WHERE ((((t0.mediapackage_id = t7.mediapackage_id) AND (t7.namespace = 'org.opencastproject.scheduler')) AND (t7.property_name = 'agent')) AND (t7.val_string = 'm320'))) );
```

The runtime of this query can be dramatically improved by including:

t0.owner='org.opencastproject.scheduler'

in the WHERE condition. Is there an easy way to do that?

krinnewitz commented:
Yes, right, this query is now the slowest part when the request is processed.

I think, if we want to check for the owner in the WHERE clause, we would have to change the AssetManager's AQueryBuilder to allow this. Unfortunately, we cannot just access the database but have several layers of abstraction (JPA + AssetManager's Query Builder) in between 😞

smarquard commented:
OK. There are some further optimizations that I think are going to be essential, but let's leave them for follow-up PRs. MH-12652 is for the restriction to owner in queries requiring the API change.

Here are my before and after testing results for this PR. These were done by merging this PR into UCT's 4.x build which is not identical to community 4.x, but largely the same (no relevant changes to this area of code). The database is a copy of our production database, which has approx 32,000 historical events in the asset manager, and started with a few hundred scheduled events.

I scheduled repeating events Mon-Fri, 1 Feb to 30 Nov, 217 total events. It took 39 mins for this to complete (before all 217 events showed up in the Admin UI). Conflict check for a single event took 8.99s, and for a group of 10 events, 1.4 minutes.

After the PR, the time for both single and group of 10 repeating events conflict check was between 7.5s and 8s. So this PR improves conflict checking from "unusable" to "just slow".

Schedule a group of 217 new events is still very slow because now the conflict check is called for each individual event (as can be seen from the mysql slow query log). However, that's a different endpoint and should be a different PR so I've filed it as MH-12651.

So this PR makes a significant improvement although further optimizations are possible.

smarquard reviewed (approved):
Code looks good and the changes produce a significant improvement.

I'm going to delay merging this for another day or so in case there are any further comments from others.

staubesv commented:
@smarquard Please wait with merging this... @krinnewitz will update this PR to see whether adding the snapshot owner can be done

Commit:
Add Check For Snapshot Owner

krinnewitz commented:
@smarquard I just added the check for the snapshot owner. My test data is not suitable for verifying if this improves the performance. If it doesn't improve the performance for you, I can revert the commit. Just let me know. It may also make sense to check if there are other queries which can be accelerated by checking for the owner, since we can now do that easily.

smarquard commented:
Thanks. I tested with this change, and the conflict time check is reduced from 7.5s to 4.5s. The SQL query no longer shows up in the slow query log so executes in < 1s on our data set. I'm guessing most of the rest of the time is spent parsing XML.

I tested this change on the query for a CA calendar (/recordings/calendars?agentid=xyz) and it reduces the query time from 4.3s to just over 1s, so is also a big win. If you can add this in, then I think we can call this PR done and merge it.

```
diff --git a/modules/matterhorn-scheduler-impl/src/main/java/org/opencastproject/scheduler/impl/SchedulerServiceImpl.java b/modules/matterhorn-scheduler-impl/src/main/java/org/opencastproject/scheduler/impl/SchedulerServiceImpl.java
index cc11502..1efba50 100644
--- a/modules/matterhorn-scheduler-impl/src/main/java/org/opencastproject/scheduler/impl/SchedulerServiceImpl.java
+++ b/modules/matterhorn-scheduler-impl/src/main/java/org/opencastproject/scheduler/impl/SchedulerServiceImpl.java
@@ -1397,7 +1397,7 @@ public class SchedulerServiceImpl extends AbstractIndexProducer implements Sched
     try {
       AQueryBuilder query = assetManager.createQuery();
       Props p = new Props(query);
-      Predicate predicate = withOrganization(query).and(query.hasPropertiesOf(p.namespace())).and(p.optOut().eq(false))
+      Predicate predicate = withOrganization(query).and(withOwner(query)).and(query.hasPropertiesOf(p.namespace())).and(p.optOut().eq(false))
               .and(withVersion(query)).and(p.end().ge(DateTime.now().minusHours(1).toDate()));
       for (String agentId : captureAgentId) {
         predicate = predicate.and(p.agent().eq(agentId));
```

Commit:
Speed up SchedulerServiceImpl.getCalendar()

By only considering snapshots with the owner
"org.opencastproject.scheduler", less rows are read when the query is
performed on the database.

krinnewitz commented:
Happy to hear this. I applied the patch you provided.
